### PR TITLE
Continuous profiling

### PR DESCRIPTION
**Pull request by christos-h**
## Motivation

To understand the performance characteristics of Linera it is useful to generate flamegraphs of the call stack.

## Proposal

Add continuous profiling using Pyroscope - a tool maintained by Grafana which samples your applications' call stack at a regular interval (say 100 Hz) and makes it queryable via a nice UI.

image

## Links

[Pyroscope](https://pyroscope.io/)
[Pyroscope Helm Chart](https://artifacthub.io/packages/helm/grafana/pyroscope)

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -86,7 +86,7 @@ RUN mv \
     target/"$target"/release/linera-db \
     ./
 
-RUN strip linera linera-proxy linera-server linera-db
+RUN strip linera linera-db
```
Comment (andresilva91) [inline docker/Dockerfile:89]: Is it worth adding a TODO here for figuring out how to use `split-debuginfo`?

File: linera-service/src/proxy.rs [hunk 2]
```diff
@@ -54,6 +54,27 @@ enum Proxy {
...
+        let agent_running = agent.start()?;
+        let proxy_res = match self {
+            Proxy::Simple(simple_proxy) => simple_proxy.run().await,
+            Proxy::Grpc(grpc_proxy) => grpc_proxy.run().await,
+        };
+        agent_running.shutdown();
+        proxy_res
+    }
+
+    /// Run the proxy with profiling.
```
Comment (andresilva91) [inline linera-service/src/proxy.rs:76]: Comments need to be swapped

File: linera-service/src/server.rs [hunk 3]
```diff
@@ -405,7 +405,7 @@ fn main() {
         .enable_all()
         .build()
         .expect("Failed to create Tokio runtime")
-        .block_on(run(options))
+        .block_on(run(options));
```
Comment (andresilva91) [inline linera-service/src/server.rs:408]: Will this make `main` always return a non error even when there's an error? 🤔

**Review by andresilva91 — COMMENTED** [review]
This is so cool 🤩 some comments/questions

File: linera-service/src/proxy.rs (hunk 2 above)
Comment (christos-h) [reply linera-service/src/proxy.rs:76]: :sweat_smile: indeed!

**Comment by christos-h** [comment]
I couldn't get this to work consistently, so closing for now.

**Pull request closed without merging**